File: src/cudgel/graph.py
"""Graph-based relationship queries."""

from typing import Any, Optional

from cudgel.config import CudgelConfig
from cudgel.database import Database


class GraphQuery:
    """Query code relationships as a graph."""

    def __init__(self, config: CudgelConfig):
        self.config = config
        self.db = Database(config)

    async def initialize(self) -> None:
        """Initialize the graph query system."""
        await self.db.connect()

    async def close(self) -> None:
        """Close connections."""
        await self.db.close()

    async def get_symbol_by_name(
        self,
        name: str,
        repository_path: Optional[str] = None
    ) -> Optional[dict[str, Any]]:
        """Get a symbol by name."""
        if not self.db.conn:
            await self.db.connect()

        async with self.db.conn.cursor() as cur:
            if repository_path:
                await cur.execute(
                    """
                    SELECT s.*, f.path, f.language, r.path as repo_path
                    FROM symbols s
                    JOIN files f ON s.file_id = f.id
                    JOIN repositories r ON f.repository_id = r.id
                    WHERE s.name = %s AND r.path = %s
                    LIMIT 1
                    """,
                    (name, repository_path),
                )
            else:
                await cur.execute(
                    """
                    SELECT s.*, f.path, f.language, r.path as repo_path
                    FROM symbols s
                    JOIN files f ON s.file_id = f.id
                    JOIN repositories r ON f.repository_id = r.id
                    WHERE s.name = %s
                    LIMIT 1
                    """,
                    (name,),
                )

            return await cur.fetchone()

    async def get_references(
        self,
        symbol_name: str,
        repository_path: Optional[str] = None,
        depth: int = 1
    ) -> dict[str, Any]:
        """
        Get reference graph for a symbol.

        Args:
            symbol_name: Name of the symbol
            repository_path: Optional filter by repository
            depth: How many levels deep to traverse

        Returns:
            Graph structure with nodes and edges
        """
        symbol = await self.get_symbol_by_name(symbol_name, repository_path)
        if not symbol:
            return {"nodes": [], "edges": [], "error": "Symbol not found"}

        nodes: dict[int, dict[str, Any]] = {}
        edges: list[dict[str, Any]] = []

        await self._traverse_references(symbol["id"], nodes, edges, depth, set())

        return {
            "nodes": list(nodes.values()),
            "edges": edges,
            "root": symbol_name,
        }

    async def _traverse_references(
        self,
        symbol_id: int,
        nodes: dict[int, dict[str, Any]],
        edges: list[dict[str, Any]],
        depth: int,
        visited: set[int],
    ) -> None:
        """Recursively traverse references."""
        if depth <= 0 or symbol_id in visited:
            return

        visited.add(symbol_id)

        if not self.db.conn:
            return

        # Get symbol info
        async with self.db.conn.cursor() as cur:
            await cur.execute(
                """
                SELECT s.*, f.path, f.language
                FROM symbols s
                JOIN files f ON s.file_id = f.id
                WHERE s.id = %s
                """,
                (symbol_id,),
            )
            symbol = await cur.fetchone()

            if symbol:
                nodes[symbol_id] = {
                    "id": symbol["id"],
                    "name": symbol["name"],
                    "kind": symbol["kind"],
                    "file": symbol["path"],
                    "language": symbol["language"],
                    "line": symbol["start_line"],
                }

            # Get outgoing references
            await cur.execute(
                """
                SELECT
                    r.*,
                    s.name as to_name,
                    s.kind as to_kind
                FROM references r
                JOIN symbols s ON r.to_symbol_id = s.id
                WHERE r.from_symbol_id = %s
                """,
                (symbol_id,),
            )
            references = await cur.fetchall()

            for ref in references:
                edges.append({
                    "from": symbol_id,
                    "to": ref["to_symbol_id"],
                    "type": ref["reference_type"],
                    "line": ref["line"],
                    "column": ref["column"],
                })

                # Recursively traverse
                await self._traverse_references(
                    ref["to_symbol_id"],
                    nodes,
                    edges,
                    depth - 1,
                    visited,
                )

    async def get_call_graph(
        self,
        symbol_name: str,
        repository_path: Optional[str] = None,
        direction: str = "outgoing"
    ) -> dict[str, Any]:
        """
        Get call graph for a function/method.

        Args:
            symbol_name: Name of the function/method
            repository_path: Optional filter by repository
            direction: "outgoing" (what this calls), "incoming" (what calls this), or "both"

        Returns:
            Call graph structure
        """
        symbol = await self.get_symbol_by_name(symbol_name, repository_path)
        if not symbol:
            return {"nodes": [], "edges": [], "error": "Symbol not found"}

        if symbol["kind"] not in ("function", "method"):
            return {"nodes": [], "edges": [], "error": "Symbol is not a function or method"}

        nodes: dict[int, dict[str, Any]] = {}
        edges: list[dict[str, Any]] = []

        if not self.db.conn:
            await self.db.connect()

        # Add root node
        nodes[symbol["id"]] = {
            "id": symbol["id"],
            "name": symbol["name"],
            "kind": symbol["kind"],
            "file": symbol["path"],
            "language": symbol["language"],
            "line": symbol["start_line"],
        }

        async with self.db.conn.cursor() as cur:
            if direction in ("outgoing", "both"):
                # Get functions this one calls
                await cur.execute(
                    """
                    SELECT
                        r.*,
                        s.id as to_id,
                        s.name as to_name,
                        s.kind as to_kind,
                        f.path as to_file
                    FROM references r
                    JOIN symbols s ON r.to_symbol_id = s.id
                    JOIN files f ON s.file_id = f.id
                    WHERE r.from_symbol_id = %s
                    AND s.kind IN ('function', 'method')
                    """,
                    (symbol["id"],),
                )
                outgoing = await cur.fetchall()

                for ref in outgoing:
                    nodes[ref["to_id"]] = {
                        "id": ref["to_id"],
                        "name": ref["to_name"],
                        "kind": ref["to_kind"],
                        "file": ref["to_file"],
                    }
                    edges.append({
                        "from": symbol["id"],
                        "to": ref["to_id"],
                        "type": "calls",
                    })

            if direction in ("incoming", "both"):
                # Get functions that call this one
                await cur.execute(
                    """
                    SELECT
                        r.*,
                        s.id as from_id,
                        s.name as from_name,
                        s.kind as from_kind,
                        f.path as from_file
                    FROM references r
                    JOIN symbols s ON r.from_symbol_id = s.id
                    JOIN files f ON s.file_id = f.id
                    WHERE r.to_symbol_id = %s
                    AND s.kind IN ('function', 'method')
                    """,
                    (symbol["id"],),
                )
                incoming = await cur.fetchall()

                for ref in incoming:
                    nodes[ref["from_id"]] = {
                        "id": ref["from_id"],
                        "name": ref["from_name"],
                        "kind": ref["from_kind"],
                        "file": ref["from_file"],
                    }
                    edges.append({
                        "from": ref["from_id"],
                        "to": symbol["id"],
                        "type": "calls",
                    })

        return {
            "nodes": list(nodes.values()),
            "edges": edges,
            "root": symbol_name,
            "direction": direction,
        }

File: src/graph.rs
//! Graph-based relationship queries

use crate::{database::Database, Result};
use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet};
use std::sync::Arc;

/// Node in a code relationship graph
///
/// Represents a symbol in the call/reference graph.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GraphNode {
    /// Symbol database ID
    pub id: i32,
    /// Symbol name
    pub name: String,
    /// Symbol kind ("function", "class", etc.)
    pub kind: String,
    /// Source file path
    pub file: String,
    /// Programming language
    pub language: Option<String>,
    /// Line number in file (1-indexed)
    pub line: i32,
}

/// Edge in a code relationship graph
///
/// Represents a relationship between two symbols (call, import, etc.).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GraphEdge {
    /// Source node ID
    pub from: i32,
    /// Target node ID
    pub to: i32,
    /// Edge type ("call", "import", "extends", etc.)
    #[serde(rename = "type")]
    pub edge_type: String,
    /// Line number where relationship occurs
    pub line: Option<i32>,
    /// Column number where relationship occurs
    pub column: Option<i32>,
}

/// Code relationship graph
///
/// Contains nodes (symbols) and edges (relationships) discovered by traversal.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Graph {
    /// All nodes in the graph
    pub nodes: Vec<GraphNode>,
    /// All edges connecting nodes
    pub edges: Vec<GraphEdge>,
    /// Root symbol name that was queried
    pub root: String,
}

/// Graph query engine
///
/// Provides methods for traversing code relationships and building call graphs.
pub struct GraphQuery {
    db: Arc<Database>,
}

impl GraphQuery {
    /// Create a new graph query engine
    ///
    /// # Arguments
    /// * `db` - Database connection
    pub fn new(db: Arc<Database>) -> Self {
        GraphQuery { db }
    }

    /// Get references graph for a symbol
    ///
    /// Traverses the code relationship graph starting from the named symbol,
    /// following outgoing references up to the specified depth.
    ///
    /// # Arguments
    /// * `symbol_name` - Name of the root symbol to start traversal from
    /// * `repository_path` - Optional filter to specific repository
    /// * `depth` - Maximum traversal depth (number of hops from root)
    ///
    /// # Returns
    /// Graph containing all discovered nodes and edges
    pub async fn get_references(
        &self,
        symbol_name: &str,
        repository_path: Option<&str>,
        depth: usize,
    ) -> Result<Graph> {
        let symbol = self
            .db
            .get_symbol_by_name(symbol_name, repository_path)
            .await?;

        let Some(symbol) = symbol else {
            return Ok(Graph {
                nodes: vec![],
                edges: vec![],
                root: symbol_name.to_string(),
            });
        };

        let symbol_id: i32 = symbol.get("id");

        let mut nodes = HashMap::new();
        let mut edges = Vec::new();
        let mut visited = HashSet::new();

        self.traverse_references(symbol_id, &mut nodes, &mut edges, depth, &mut visited)
            .await?;

        Ok(Graph {
            nodes: nodes.into_values().collect(),
            edges,
            root: symbol_name.to_string(),
        })
    }

    async fn traverse_references(
        &self,
        symbol_id: i32,
        nodes: &mut HashMap<i32, GraphNode>,
        edges: &mut Vec<GraphEdge>,
        depth: usize,
        visited: &mut HashSet<i32>,
    ) -> Result<()> {
        if depth == 0 || visited.contains(&symbol_id) {
            return Ok(());
        }

        visited.insert(symbol_id);

        // Get symbol information and add to nodes
        let symbol_row = self.db.get_symbol_by_id(symbol_id).await?;
        if let Some(row) = symbol_row {
            nodes.entry(symbol_id).or_insert_with(|| GraphNode {
                id: symbol_id,
                name: row.get("name"),
                kind: row.get("kind"),
                file: row.get("path"),
                language: row.get("language"),
                line: row.get("start_line"),
            });
        }

        let references = self.db.get_references(symbol_id).await?;

        for reference in references {
            edges.push(GraphEdge {
                from: reference.from_symbol_id,
                to: reference.to_symbol_id,
                edge_type: reference.reference_type,
                line: Some(reference.line),
                column: Some(reference.column),
            });

            // Get the target symbol and add to nodes
            let to_symbol_row = self.db.get_symbol_by_id(reference.to_symbol_id).await?;
            if let Some(row) = to_symbol_row {
                nodes
                    .entry(reference.to_symbol_id)
                    .or_insert_with(|| GraphNode {
                        id: reference.to_symbol_id,
                        name: row.get("name"),
                        kind: row.get("kind"),
                        file: row.get("path"),
                        language: row.get("language"),
                        line: row.get("start_line"),
                    });
            }

            // Recursively traverse
            Box::pin(self.traverse_references(
                reference.to_symbol_id,
                nodes,
                edges,
                depth - 1,
                visited,
            ))
            .await?;
        }

        Ok(())
    }

    /// Get call graph for a symbol
    ///
    /// Similar to `get_references` but filtered for function call relationships.
    /// Currently delegates to `get_references` with depth=2.
    ///
    /// # Arguments
    /// * `symbol_name` - Name of the root symbol
    /// * `repository_path` - Optional filter to specific repository
    /// * `_direction` - Direction of traversal (not yet implemented)
    ///
    /// # Returns
    /// Graph containing call relationships
    pub async fn get_call_graph(
        &self,
        symbol_name: &str,
        repository_path: Option<&str>,
        _direction: &str,
    ) -> Result<Graph> {
        // Similar to get_references but filtered for function calls
        self.get_references(symbol_name, repository_path, 2).await
    }
}
